File: C++/BinaryTreeLevelOrderTraversal.cpp
#include<iostream>
#include<vector>
#include<queue>
#include<string>
#include<string.h>
#include<math.h>
#include<algorithm>
using namespace std;
struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
	vector<vector<int>> ret;
	vector<vector<int> > levelOrder(TreeNode *root) {
		// Start typing your C/C++ solution below
		// DO NOT write int main() function
		ret.clear();
		if(!root)
			return ret;
		queue<TreeNode*> q;
		q.push(root);
		int pre = 1;
		int next = 0;
		vector<int> list;
		while(!q.empty())
		{
			TreeNode* node = q.front();
			q.pop();
			pre--;
			list.push_back(node->val);
			if(node->left)
			{
				q.push(node->left);
				next++;
			}
			if(node->right)
			{
				q.push(node->right);
				next++;
			}
			if(pre == 0)
			{
				pre = next;
				next = 0;
				ret.push_back(list);
				list.clear();
			}
		}
		return ret;
	}
};

File: JAVA/BinaryTreeLevelOrderTraversal.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;

class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;

	TreeNode(int x) {
		val = x;
	}
}

class Solution {
	ArrayList<ArrayList<Integer>> ret = new ArrayList<ArrayList<Integer>>();
	public ArrayList<ArrayList<Integer>> levelOrder(TreeNode root) {
		// Start typing your Java solution below
		// DO NOT write main() function
		ret.clear();
		if(root == null)
			return ret;
		Queue<TreeNode> queue = new LinkedList<TreeNode>();
		queue.add(root);
		int curLevelCount = 1;
		int nextLevelCount = 0;
		ArrayList<Integer> list = new ArrayList<Integer>();
		while(!queue.isEmpty())
		{
			curLevelCount--;
			TreeNode node = queue.poll();
			list.add(node.val);
			if(node.left != null)
			{
				queue.add(node.left);
				nextLevelCount++;
			}
			if(node.right != null)
			{
				queue.add(node.right);
				nextLevelCount++;
			}
			if(curLevelCount == 0)
			{
				curLevelCount = nextLevelCount;
				nextLevelCount = 0;
				ret.add(new ArrayList<Integer>(list));
				list.clear();
			}
		}
		return ret;
	}
}

public class BinaryTreeLevelOrderTraversal {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		// TODO Auto-generated method stub

	}

}
